language-support/java/bindings/src/main/java/com/daml/ledger/javaapi/data/codegen/json/JsonLfEncoders.java

```diff
@@ -0,0 +1,202 @@
+    return w -> w.write(String.valueOf(value));
+  }
+
+  public static JsonLfEncoder int64(Long value) {
+    return w -> w.write(String.valueOf(value));
+  }
+
+  public static JsonLfEncoder text(String value) {
+    return w -> {
+      String escaped = String.valueOf(stringEncoder.quoteAsString(value));
+      w.write("\"" + escaped + "\"");
+    };
+  }
+
+  public static JsonLfEncoder numeric(BigDecimal value) {
+    return w -> w.write(value.toString());
+  }
+
+  public static JsonLfEncoder timestamp(Instant t) {
+    return w -> {
+      String subsecPat = isRoundTo(t, SECONDS) ? "" : (isRoundTo(t, MILLIS) ? ".SSS" : ".SSSSSS");
+      var formatter =
+          DateTimeFormatter.ofPattern("yyyy-MM-dd'T'HH:mm:ss" + subsecPat + "X")
+              .withZone(ZoneOffset.UTC);
+      w.write("\"" + formatter.format(t) + "\"");
+    };
+  }
+
+  public static JsonLfEncoder date(LocalDate d) {
+    return text(d.toString());
+  }
+
+  public static JsonLfEncoder party(String value) {
+    return text(value);
+  }
+
+  public static JsonLfEncoder contractId(ContractId<?> value) {
+    return text(value.toValue().asContractId().get().getValue());
+  }
+
+  public static <E extends Enum<E>> Function<E, JsonLfEncoder> enumeration(
+      Function<E, String> toDamlName) {
+    return value -> text(toDamlName.apply(value));
+  }
+
+  public static <T> Function<Iterable<T>, JsonLfEncoder> list(
+      Function<T, JsonLfEncoder> itemEncoder) {
+    return items -> {
+      return w -> {
+        w.writeStartArray();
+        boolean first = true;
+        for (var item : items) {
+          if (!first) w.writeComma();
+          itemEncoder.apply(item).encode(w);
+          first = false;
+        }
+        w.writeEndArray();
+      };
+    };
+  }
+
+  public static <T> Function<Map<String, T>, JsonLfEncoder> textMap(
+      Function<T, JsonLfEncoder> valueEncoder) {
+    return map -> {
+      return w -> {
+        w.writeStartObject();
+        boolean first = true;
+        for (var entry : map.entrySet()) {
+          if (!first) w.writeComma();
+          w.writeFieldName(entry.getKey());
+          valueEncoder.apply(entry.getValue()).encode(w);
+          first = false;
+        }
+        w.writeEndObject();
+      };
+    };
+  }
+
+  public static <K, V> Function<Map<K, V>, JsonLfEncoder> genMap(
+      Function<K, JsonLfEncoder> keyEncoder, Function<V, JsonLfEncoder> valueEncoder) {
+    return map -> {
+      return w -> {
+        w.writeStartArray();
+        boolean first = true;
+        for (var entry : map.entrySet()) {
+          if (!first) w.writeComma();
+          w.writeStartArray();
+          keyEncoder.apply(entry.getKey()).encode(w);
+          w.writeComma();
+          valueEncoder.apply(entry.getValue()).encode(w);
+          w.writeEndArray();
+          first = false;
+        }
+        w.writeEndArray();
+      };
+    };
+  }
+
+  public static <T> Function<Optional<T>, JsonLfEncoder> optional(
```

Comment:
consider adding an assertion that T is not Optional.

Reply:
Done